### ISSUE TITLE
dub test shows insufficient information when encountering an invalid file

### ISSUE THREAD

**Issue by schveiguy** [dlang/dub#1922]
### System information
- **dub version**: latest
- **OS Platform and distribution**: all
- **compiler version**: all

### Bug Description
`dub test` attempts to parse all `*.d` files in the source directory for information. If any files have invalid utf, the `std.file.readText` function throws an exception, for which dub just prints the exception message and exits.

See log for example.

### How to reproduce?

- Create a blank dub project
- Create a file that is not a valid utf-8 file, that ends in a d extension in the source directory.
- Run `dub test`

### Expected Behavior
Error that shows what file is causing the issue.

### Logs
```
Generating test runner configuration 'testdub-test-library' for 'library' (library).
Invalid UTF-8 sequence (at index 1)
```

**Comment by schveiguy** [comment]
Note, the use case I had was related to #1372, but I first encountered using dub test. Mac seems to create a shadow directory for every directory on a server called `.AppleDouble` with all the files repeated under that directory with the same name. So `app.d` has a shadow `.AppleDoublel/app.d`. This shadow file is not a utf-8 file.

Encountering this error, and not knowing the hidden directories were there, it took me a long time to figure out why this error was happening.

**Comment by rikkimax** [comment]
In case anybody else is wondering, that error message is indeed not generated by dmd-fe.
So this is a dub bug.

**Comment by schveiguy** [comment]
I should have posted this earlier, because I already diagnosed it using a debugger. It's generated by `std.file`. The offending line is here: https://github.com/dlang/dub/blob/4ed2091d04952c46a2a02b467a6ae08689e7bf5d/source/dub/internal/utils.d#L677

Or rather, that's the place where the error is occurring. I don't know who is catching the error and printing the message. I don't think it is druntime, because that should print a full stack trace (which would have been super-useful!)

The actual exception being thrown from the bowels of `std.utf`. Hard to tell where, it's a maze.

**Comment by schveiguy** [comment]
lo and behold, the reason the stack trace isn't printed isn't dub's fault. It's `UTFException`, which is overriding `Object.toString`, but not printing the stack trace. If you print the exception with just `writeln(e)`, then it prints the stack trace, because it uses the `toString` overload with a sink. UTFException should be updated (as well as we should be clearer here what happened).

**Comment by WebFreak001** [comment]
dub catches all exceptions when running as application at some point, use `-v` to print full exception stacktrace.

I think it would make sense to add some code like `scope (failure) logError("Failed to read from file %s.", filePath);` at some point where it doesn't try to recover

**Comment by schveiguy** [comment]
> use `-v` to print full exception stacktrace

Nope. Doesn't work. But that's not dub's fault exactly. See https://issues.dlang.org/show_bug.cgi?id=20735

**Comment by schveiguy** [comment]
I think the right answer is to throw an exception containing the file name, and have that print the stack trace.
